frontend/catalyst/qfunc.py

```diff
@@ -106,3 +120,110 @@ def _eval_quantum(*args):
         args_flat = tree_flatten(args)[0]
         res_flat = func_p.bind(flattened_fun, *args_flat, fn=self)
         return tree_unflatten(out_tree_promise(), res_flat)[0]
+
+
+# pylint: disable=protected-access,no-member,not-callable
+def dynamic_one_shot(qnode):
+    """Transform a QNode to into several one-shot tapes to support dynamic circuit execution.
+
+    Args:
+        qnode (QNode): a quantum circuit which will run ``num_shots`` times
+
+    Returns:
+        qnode (QNode):
+
+        The transformed circuit to be run ``num_shots`` times such as to simulate dynamic execution.
+
+
+    **Example**
+
+    Consider the following circuit:
+
+    .. code-block:: python
+
+        dev = qml.device("lightning.qubit", shots=100)
+        params = np.pi / 4 * np.ones(2)
+
+        @qjit
+        @dynamic_one_shot
+        @qml.qnode(dev, diff_method=None)
+        def circuit(x, y):
+            qml.RX(x, wires=0)
+            m0 = measure(0, reset=reset, postselect=postselect)
+
+            @cond(m0 == 1)
+            def ansatz():
+                qml.RY(y, wires=1)
+
+            ansatz()
+            return measure_f(wires=[0, 1])
+
+    The ``dynamic_one_shot`` decorator prompts the QNode to perform a hundred one-shot
+    calculations, where in each calculation the ``measure`` operations dynamically
+    measures the 0-wire and collapse the state vector stochastically.
+    """
+
+    cpy_tape = None
+    aux_tapes = None
+
+    def transform_to_single_shot(qnode):
+        nonlocal cpy_tape
+        nonlocal aux_tapes
+
+        @qml.transform
+        def dynamic_one_shot_partial(
```

Comment:
How do we want to handle users specifying `mcm_method="one-shot"` and also decorating with `dynamic_one_shot`? In the PL docs currently we have a warning saying that doing that could lead to unexpected behaviour, but there isn't any validation blocking users from doing so. Do we want to keep it that way?

If not, it would be helpful for my work to have this transform outside the `dynamic_one_shot` function. so that I can check if it already exists in the qnode's transform program.

Reply:
> How do we want to handle users specifying mcm_method="one-shot" and also decorating with dynamic_one_shot? In the PL docs currently we have a warning saying that doing that could lead to unexpected behaviour, but there isn't any validation blocking users from doing so. Do we want to keep it that way?

This is one of the big reasons I think there should only be one way to set the execution mode.